Add wallet interface for server-initiated recurring payments

PetrDlouhy opened the pull request:
Changes of Payment model, that are required to implement PayU backend and recurring payments.

I could implement these changes through some mixin mandatory only for PayU backend, but I think, that those methods might be common to more backends and it would be highly usefull to have same implementation for all backends. I am willing to rewrite them bit more general shape, if requested.

patrys commented on payments/models.py:
```diff
@@ -133,6 +133,68 @@ def get_success_url(self):
+        """ Get the user asociated with this payment """
+        raise NotImplementedError()
+
+    def get_user_email(self):
+        """ Get user email """
+        try:
+            return self.get_user().email
+        except AttributeError:
+            return None
+
+    def get_user_first_name(self):
+        """
+        Get user first name
+        Used only by PayU provider for now
+        """
+        try:
+            return self.get_user().first_name
+        except AttributeError:
+            return None
+
+    def get_user_last_name(self):
+        """
+        Get user last name
+        Used only by PayU provider for now
+        """
+        try:
+            return self.get_user().last_name
+        except AttributeError:
+            return None
```
This ties the package to a particular shape of the user object, it would not work in at least some of our projects. Could those fields not be extracted from billing information?

PetrDlouhy replied:
These are the implementations for the most common case with intention, that users will override it in any other cases.

The billing information is info about the merchant, not the user, which is required by PayU (or PayU will ask that information in next step, which degrades UX)

PetrDlouhy replied:
But maybe it would be appropriate to throw `NotImplementedError` instead of returning None and write some note to the comments about thet.

PetrDlouhy commented:
Hi @patrys, I rebased this to current master, and the methods are throwing `NotImplementedError`, if they can't figure out the variables from user.

Can you please give a new review? I think, it would be handy to unify these methods among providers.

WhyNotHugo commented on payments/models.py:
```diff
@@ -135,6 +135,71 @@ def get_success_url(self) -> str:
     def get_process_url(self) -> str:
         return reverse('process_payment', kwargs={'token': self.token})
 
+    def get_payment_url(self):
```
You can raise `RedirectNeeded` in you provider code. Any application using this lib should catch these and redirect as needed.

Is there a reason that won't work for you?

WhyNotHugo commented on payments/models.py:
```diff
@@ -135,6 +135,71 @@ def get_success_url(self) -> str:
+        """
+        raise NotImplementedError()
+
+    def get_user(self):
```
What's this supposed to return?

PetrDlouhy replied:
This is simple helper method used by the other `get_user_*` methods, so it is the only method, that needs to be overridden in case of the Django `django.contrib.auth.models.User` model.

WhyNotHugo replied:
You're assuming that all applications have a `User` for each payment made.

I'm not a fan of imposing such a big restriction, especially without the need for one. You probably want a `user` in your own `Payment` instance, not in the abstract one in this package.

WhyNotHugo commented on payments/models.py:
```diff
@@ -135,6 +135,71 @@ def get_success_url(self) -> str:
+        """ Get the user asociated with this payment """
+        raise NotImplementedError()
+
+    def get_user_email(self):
```
What's wrong with `billing_email`? What's expected of this method?

PetrDlouhy replied:
The `billing_*` fields hold information about the issuer (Me), PayU (and potentially other providers) needs information about the user.

WhyNotHugo replied:
The `billing_` fields all hold billing details, that is: the details of the person paying (this is what payment providers usually ask for, and the thing that actually varies on a per-payment basis).

This is the assumption made for other payment implementations, so it's the assumption that any third-party providers should make two.

Using `billing_` fields as something elsemeans that applications cannot use PayU **and** another provider, since the give different semantic meaning to the same fields.

PetrDlouhy replied:
Oh, this is something, I got totally wrong. Sorry for loss of your time and thank you for clearing that out.
I used this for storing issuer info when I was implementing PayPal (which doesn't use it at all), and then continued with the wrong implementation for PayU and recurring payments.

Maybe it would be useful to put some info about those fields into docs, I will try to write something.

WhyNotHugo commented on payments/models.py:
```diff
@@ -135,6 +135,71 @@ def get_success_url(self) -> str:
+        except AttributeError:
+            raise NotImplementedError()
+
+    def get_user_first_name(self):
```
Ditto `billing_first_name`.

WhyNotHugo commented on payments/models.py:
```diff
@@ -135,6 +135,71 @@ def get_success_url(self) -> str:
+        except AttributeError:
+            raise NotImplementedError()
+
+    def get_user_last_name(self):
```
Ditto `billing_last_name`.

WhyNotHugo commented on payments/models.py:
```diff
@@ -135,6 +135,71 @@ def get_success_url(self) -> str:
+        except AttributeError:
+            raise NotImplementedError()
+
+    def get_renew_token(self):
```
Since this needs to be implemented by an application developer: what's expected here?
Isn't this something internal to the provider?

PetrDlouhy replied:
To make recurring payments work, there needs to be token associate with the user.
I wanted to provide standardized way to store textual tokens for usage across `django-payments` payment providers.

WhyNotHugo replied:
You can use `payment.attrs` to store custom per-provider data. I'd suggest using that.

For now it's a proxy that JSON-encodes into a string field (legacy reasons). I'm planning to migrate this into a plain `JSONField`, but usage should remain pretty much the same.

WhyNotHugo replied:
An alternative would also be for you to create an explicit `Subscription` object, and save the data there.

You can then link payments to subscriptions. I think this also slightly mirrors how payment providers themselves model these things.

PetrDlouhy replied:
The `Subscription` model is similar to what I am storing the parameters of this method to:
https://github.com/django-getpaid/django-plans/blob/master/plans/models.py#L402

So the abstraction could look something like this:
```python
class BaseSubscription(models.Model):
    user_plan = models.OneToOneField('UserPlan', on_delete=models.CASCADE, related_name='recurring')
    token = models.CharField(
        _('recurring token'),
        help_text=_('Token, that will be used for payment renewal. Depends on used payment provider'),
        max_length=255,
        default=None,
        null=True,
        blank=True,
    )
    card_expire_year = models.IntegerField(null=True, blank=True)
    card_expire_month = models.IntegerField(null=True, blank=True)
    card_masked_number = models.CharField(null=True, blank=True, max_length=255)
        
    class Meta: 
        abstract = True

```

But I am not sure, how to implement this to `django-payments` without implying too much about the usage logic e.g. how to initialise such objects inside `Payment` model.

WhyNotHugo replied:
Probably the concrete `Payment` class created by users of `django-plans` would have to include a field specific for it. You can also provide a mixin, so that implementors do something like

```python
class Payment(BasePayment, SubscriptionMixin):
    # SubscriptionMixin has a foreign key to the concrete subscription model.
```

I do have to point out that `card_expire_year`, `card_expire_month` and `card_masked_number` all make the assumption that the payment method was a credit card, which is not always the case.

WhyNotHugo reviewed (commented):
I've a few questions to further understand the needs behind this.

PetrDlouhy commented:
Here is sample implementation of the methods that connects `django-payments` to `django-plans`: https://github.com/PetrDlouhy/django-plans-payments/blob/feature/recurring-payments/plans_payments/models.py

WhyNotHugo reviewed (commented):
Apologies for the delay, each time I re-reviewed this I had different questions.

I think this looks good, I only have a few doubts on the current implementation...

PetrDlouhy commented on payments/models.py:
```diff
@@ -36,6 +38,60 @@ def __setattr__(self, key, value):
 
 
+class BaseSubscription(models.Model):
+    token = models.CharField(
+        _("subscription token/id"),
+        help_text=_("Token/id used to identify subscription by provider"),
+        max_length=255,
+        default=None,
+        null=True,
+        blank=True,
+    )
```
I think, that some kind of token or payment identification would be needed for most implementation.
I used this is for Subscription ID in case of PayPal and for Card token in case of PayU. Accessing that through JSONField would be quite unflexible.

But adding JSONField for additional fields might be a good idea.

WhyNotHugo replied:
> But adding JSONField for additional fields might be a good idea. 

Do you want to add those on this PR, or follow-up later?

PetrDlouhy commented on payments/models.py:
```diff
@@ -36,6 +38,60 @@ def __setattr__(self, key, value):
+        month = "month"
+        day = "day"
+
+    def get_token(self) -> str:
+        return self.token
```
I wanted to add a level of encapsulation for potential overrides. But it is probably really redundant.

PetrDlouhy commented on payments/models.py:
```diff
@@ -144,6 +200,33 @@ def get_success_url(self) -> str:
+        """
+        raise NotImplementedError()
+
+    def get_subscription(self) -> Optional[BaseSubscription]:
+        """
+        Returns subscription object associated with this payment
+        or None if the payment is not recurring
+        """
+        return None
```
I think this enables more loose implementation.

I am using `django-plans` to which I added the recurring functionality with RecurringUserPlan model. Now I can do it without any ties to `django-payments` (it works even if the model is not based on BaseSubscription, it just has the same fields).

I am not sure, what would the foreign key would bring.

WhyNotHugo replied:
Agreed.

radekholy24 replied:
I would go even further and change the `autocomplete_with_subscription`'s signature to `autocomplete_with_subscription(self, subscription)`. Then we would not need `get_subscription` (and `is_recurring`) at all and enable implementations to couple payments and subscriptions any way they want.

WhyNotHugo commented on payments/models.py:
```diff
@@ -36,6 +38,57 @@ def __setattr__(self, key, value):
+        month = "month"
+        day = "day"
+
+    def set_recurrence(self, token: str, **kwargs):
+        """
+        Sets token and other values associated with subscription recurrence
+        Kwargs can contain provider-specific values
+        """
+        self.token = token
```
I guess `kwargs` here would be assigned to that JSONField...?

WhyNotHugo reviewed (approved):
Makes sense to me.

PetrDlouhy commented:
I have rebased this PR and fixed testing. I also added the requested `JSONField`.

@WhyNotHugo @patrys Could you please take a look? What needs to be done to make this merged?

radekholy24 commented on payments/core.py:
```diff
@@ -119,6 +119,22 @@ def get_return_url(self, payment, extra_data=None):
+
+        Throws RedirectNeeded if there is problem with the payment that needs to be solved by user
+        """
+        raise NotImplementedError()
```
+1

It took me several hours to understand what the function is supposed to do. I guess that django-payments rather refers to this as to "making a payment". I would even not hesitate to call it `pay_with_subscription()` (ditto for the `BasePayment` class).

To clarify the docstring even more, I would say that this method is used by _payments_ not _providers_ and that it is used in an _application_-initiated workflow rather than in _server_-initiated workflow as it is not clear what server is meant (I believe I have seen somewhere that django-payments refers to the users of this library as to applications).

radekholy24 commented on payments/models.py:
```diff
@@ -39,6 +40,63 @@ def __setattr__(self, key, value):
+        null=True,
+        blank=True,
+    )
+    payment_provider = models.CharField(
+        _("payment provider"),
+        help_text=_("Provider variant, that will be used for payment renewal"),
+        max_length=255,
+        default=None,
+        null=True,
+        blank=True,
+    )
```
Why not to call it `variant` (or `payment_variant`) as in the BasePayment? Is there a particular reason for this inconsistency? I mean, I do not have any good reason for having this consistent but `provider` usually refers to the particular instance of a Provider so it might be confusing...

radekholy24 commented on payments/models.py:
```diff
@@ -39,6 +40,63 @@ def __setattr__(self, key, value):
+        null=True,
+        blank=True,
+    )
+    payment_provider = models.CharField(
+        _("payment provider"),
+        help_text=_("Provider variant, that will be used for payment renewal"),
+        max_length=255,
+        default=None,
+        null=True,
+        blank=True,
+    )
```
Also, is there any particular reason for allowing blank/null values? I mean, allowing this usually implies a lot of tedious `None`-checking in the code (which you do not do).

PetrDlouhy replied:
Here probably no.

radekholy24 commented on payments/models.py:
```diff
@@ -39,6 +40,63 @@ def __setattr__(self, key, value):
+        month = "month"
+        day = "day"
+
+    def set_recurrence(self, token: str, **kwargs):
+        """
+        Sets token and other values associated with subscription recurrence
+        Kwargs can contain provider-specific values
+        """
+        self.token = token
+        self.subscribtion_data = kwargs
```
Any particular reason for using this method if one can modify the attributes directly? I mean, if yes, would it make sense to make the attributes private? Also, if yes, I am not sure if some extra care regarding (un)intentional merging/overwriting `subscription_data` should not be provided... I mean, I would bet that sooner or later users will complain about this being too aggressive or not enough flexible...

radekholy24 commented on payments/models.py:
```diff
@@ -39,6 +40,63 @@ def __setattr__(self, key, value):
+
+    def cancel(self):
+        """
+        Cancel the subscription by provider
+        Used by providers, that use provider initiated subscription workflow
+        Implementer is responsible for cancelling the subscription model
```
Could you please clarify this docstring a bit? I am not sure if it really says what it is supposed to do. I mean, what do you mean by:

1. Cancel the subscription **by provider** - is there any "by <something>" needed? Isn't "cancel the subscription" enough?
2. **Used by providers**, that use provider initiated subscription workflow - cannot it potentially be used by anyone, if the provider supports it?
3. **Implementer is responsible for cancelling the subscription model**

radekholy24 commented on payments/models.py:
```diff
@@ -39,6 +40,63 @@ def __setattr__(self, key, value):
+    def get_unit(self) -> TimeUnit:
+        raise NotImplementedError
+
+    def cancel(self):
```
Wouldn't it be handy to have a `status` field indicating whether the subscription is pending/active/canceled or whatever the payment gateways support?

radekholy24 commented on payments/models.py:
```diff
@@ -39,6 +40,63 @@ def __setattr__(self, key, value):
 
 
+class BaseSubscription(models.Model):
+    token = models.CharField(
+        _("subscription token/id"),
+        help_text=_("Token/id used to identify subscription by provider"),
+        max_length=255,
+        default=None,
+        null=True,
+        blank=True,
+    )
```
Any particular reason for allowing blank/null values? I mean, allowing this usually implies a lot of tedious None-checking in the code.

PetrDlouhy replied:
But could we expect, that all providers will have tokens/IDs?

radekholy24 commented on payments/models.py:
```diff
@@ -39,6 +40,63 @@ def __setattr__(self, key, value):
+        self.token = token
+        self.subscribtion_data = kwargs
+
+    def get_period(self) -> int:
+        raise NotImplementedError
+
+    def get_unit(self) -> TimeUnit:
+        raise NotImplementedError
```
What about leaving this up to the subclasses? I mean, I can imagine that this interface might be too flexible for someone (e.g. PayU uses only days so it might be ambiguous how to convert from e.g. years to days) and not enough flexible for others (irregular subscriptions, minute-based subscriptions, day-in-month subscriptions, ...)

radekholy24 reviewed (approved):
Hi, here is my review as you requested...

I also added comments to a couple of existing threads:
https://github.com/jazzband/django-payments/pull/217#discussion_r1579367922
https://github.com/jazzband/django-payments/pull/217#discussion_r1579396729

Also, I expect that the documentation and tests are going to be added soon.

All of my comments are things to consider rather than change requests. Looks good to me; approving then.

radekholy24 commented on payments/models.py:
```diff
@@ -185,6 +243,39 @@ def get_success_url(self) -> str:
+    def is_recurring(self) -> bool:
+        return self.get_subscription() is not None
+
+    def autocomplete_with_subscription(self):
+        """
+        Complete the payment with subscription
+
+        If the provider uses workflow such that the payments are initiated from
+        implementer's side.
+        Call this function right before the subscription end to
+        make a new subscription payment.
+
+        Throws RedirectNeeded if there is problem with the payment
+        that needs to be solved by user
+        """
+        provider = provider_factory(self.variant)
+        provider.autocomplete_with_subscription(self)
```
Oh, and... Would it make sense to check if the payment has an allowed `status` (`WAITING`) only?

PetrDlouhy commented:
After long discussion with @radekholy24 we realized, that this interface might not be used only for Subscriptions. The logic of PayU and other providers with server initiated payments offers possibility to make subsequent payments regardless of time frame.
I think it should be rather called a Walllet (other suggestions?).

note: This is in opposition of provider initiated payments, like PayPal, where it probably IS a subscription. But I think, that logic is too different to be implemented in one interface.

If we are implementing a Wallet, then we can simplify the logic, because we don't have to deal with time units, periods etc.

Then the workflow will be following:

1. Implementator of `django-payments` creates a payment and calls `Payment.autocomplete_with_wallet()`.
2. The function `Provider.autocomplete_with_wallet(payment)` is called
3. The provider will implement the logic, complete the payment and set up `BaseWallet.token` and `BaseWallet.extra_data` and `BaseWallet.status`
4. The provider will call `BaseWallet.payment_completed(payment)`
5. Implementator can override `BaseWallet.payment_completed` and place own logic there

The basic models are:

```
BaseWallet:
    - token
    - status
    - extra_data
    - erase()
    - payment_completed()
    
Payment:
   - autocomplete_with_wallet()
```

@patrys @WhyNotHugo What do you think about this change? Any ideas?

PetrDlouhy commented:
Now we changed the logic to the wallet logic described above.
@WhyNotHugo @patrys @auvipy Any thoughts?

patrys commented:
Please don't tag me, I have not worked on this package in years

PetrDlouhy commented:
@WhyNotHugo The Wallet model represents type of recurring payments model used by PayU and other payment providers.
In this model the provider let the service provider charge the users directly with various charges. This is opposed to the model used by PayPal where the recurring subscription is managed by PayPal which only sends periodical confirmations that the payment succeeded.

PetrDlouhy commented:
I have significantly improved this PR in past two weeks. I added more documentation, updated the interface a bit etc.
This was all motivated by recurring Stripe payments implementation in PR #467. I am still working on that PR and I would like to wait until I have some more experience with recurring Stripe payments from my server. But I think this one (the interface) is ready to be merged

Now this interface is somehow confirmed to be viable by 3 providers: PayU, Stripe and Dummy.

@WhyNotHugo What do you think now. Can we proceed with this or are there any more issues to be addressed?

PetrDlouhy committed:
Add wallet interface for server-initiated recurring payments

A wallet is a stored, chargeable payment method: after one checkout in
which the user consents to future charges, the provider returns a token
(card token at PayU, vault payment token at PayPal, PaymentMethod id at
Stripe) that lets the merchant charge the user again server-side, with
no user interaction (merchant-initiated transactions, "card on file").
The application decides when and how much to charge: the amount is
taken from payment.total and can differ on every charge.

The interface has two layers:

- The contract: BasePayment.autocomplete_with_wallet() dispatching to
  BasicProvider.autocomplete_with_wallet(payment), the storage hooks
  get_renew_token() / get_renew_data() / set_renew_token(token, **kwargs),
  and BasicProvider.erase_wallet(token) for revoking the stored method.
  Providers raise RedirectNeeded when a charge turns out to need user
  interaction after all (3-D Secure / CVV re-verification).

- BaseWallet: an optional abstract model implementing the storage half
  with a PENDING -> ACTIVE -> ERASED lifecycle (a token only becomes
  chargeable after the first successful payment, and never again after
  erasure). Integrators with their own token storage use the hooks only.

The shape is extracted from three independent implementations rather
than designed up front: django-payments-payu (released), a PayPal
Complete Payments (Orders v2 + Vault) provider running in production,
and Stripe (#467). DummyProvider carries the reference implementation;
a mock-based contract suite and an ORM-based testapp suite cover it.

Compared to earlier revisions of this PR, provider-managed subscriptions
(BaseSubscription and friends) are dropped entirely: they are a
different recurring-payments paradigm with no consumers, and bundling
the two concepts was the main source of review confusion.

All additions are backward compatible: the hook defaults are no-ops and
existing providers keep working unchanged.

PetrDlouhy commented:
@WhyNotHugo Thanks for the patience with this one - I've reworked the PR to
answer your question directly, and cut it roughly in half.

**"What does `BaseWallet` represent?"** - A *wallet* is a stored, chargeable
payment method: after one checkout where the user consents to future charges,
the provider returns a token (card token at PayU, vault payment token at
PayPal, PaymentMethod id at Stripe) that lets the merchant charge the user
again server-side, without any user interaction - what the payments industry
calls *merchant-initiated transactions* / "card on file". The rewritten
`docs/wallet.rst` now opens with exactly this definition.

**What changed in this push:**

- **Removed the provider-managed subscription half entirely**
  (`BaseSubscription`, `autocomplete_with_subscription`,
  `cancel_subscription`, their tests and migration). It had no consumers and
  it was conflating two different recurring-payment paradigms in one PR -
  I believe that conflation was the main source of confusion. What remains
  is one concept only.
- **The interface is now presented in two explicit layers**: (1) the
  contract - provider methods (`autocomplete_with_wallet`,
  `erase_wallet(token)`) and `BasePayment` hooks (`get_renew_token` /
  `get_renew_data` / `set_renew_token`) - and (2) `BaseWallet` as an
  *optional* ready-made storage implementing the token lifecycle
  (PENDING → ACTIVE → ERASED). Integrators with their own storage use only
  layer 1.
- `erase_wallet` now takes the token (all known providers only need the
  token to revoke: PayU card-token delete, PayPal vault-token DELETE,
  Stripe PaymentMethod detach); wallet bookkeeping stays on the model.
- Docs rewritten from 569 lines to ~220, rebased onto current `main`.

**Why this shape - it's extracted from real implementations, not designed
up front:** [django-payments-payu](https://github.com/PetrDlouhy/django-payments-payu)
(released) implements it for PayU; a PayPal Complete Payments (Orders v2 +
Vault) provider implementing the same contract is running in production at
Blendkit and is intended as an upstream PR here once this interface lands;
Stripe (#467) implements it too (parked for now). Three providers with very
different gateways converged on these functions - that's the interface this
PR standardizes.

One open question: if the "wallet" naming still reads wrong to you, I'm open
to renaming (e.g. "stored payment method") - though the current names are in
released consumer packages, so I'd prefer keeping them unless you feel
strongly.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

PetrDlouhy commented:
I have added a draft PR with a PayPal PPCP implementation, including recurring payments, in #490.

In December 2025 I parked my effort to implement the Stripe provider (their fees are too high for us) and I am now planning to  move the PayPal integration in our stack to PPCP instead.

This means PPCP is the third provider implementing this interface, which I take as confirmation that the wallet approach is a viable common denominator for recurring payments across providers — even PayPal is steering integrations away from its IPN workflow, and its modern platform does recurring payments exactly this way (vaulted payment methods charged by the merchant).
